### ISSUE TITLE
tests fail on QNX

### ISSUE THREAD

**Issue by minyang** [nanomsg/nanomsg#433]
nanomsg version: 0.5-beta-97-ga532b66
QNX: 6.5.0

Part of the test suite fail as follows (`make check`):
- **tests/ipc**: hangs at the first call to `test_send()` of the ping-pong test<pre>
  test_send (sc, "0123456789012345678901234567890123456789");
  </pre>
  which calls, and hangs at, `rc = nn_send (sock, data, data_len, 0);` of `test_send_impl (char *file, int line, int sock, char *data)` in `testutil.h`
- **tests/ipc_shutdown**:<pre>
  Assertion failed: 3 == NN_TIMER_STATE_IDLE (src/aio/timer.c:62)
  Abort (core dumped) 
  FAIL tests/separation (exit status: 134)
  </pre>
- **tests/separation**:<pre>
  Assertion failed: 3 == NN_TIMER_STATE_IDLE (src/aio/timer.c:62)
  Abort (core dumped) 
  FAIL tests/ipc_shutdown (exit status: 134)
  </pre>

**Comment by gdamore** [comment]
Does QNX have IPC in the form of UNIX domain sockets?  I'd be pretty unsurprised to learn that it doesn't work like you'd think it should...

**Comment by gdamore** [comment]
Well, the claim from QNX.com is that it *should* work.

Sadly, QNX isn't free, so I'm not sure that I'm prepared to invest a lot of effort trying to fix any problems that might occur on that platform.

**Comment by minyang** [comment]
QNX has [free 30-day free evaluation program](http://www.qnx.com/products/evaluation), although I'm not sure if nanomsg has any plan to officially support QNX.  I'm actively using nanomsg on QNX and it has been working great.  My use case is rather simple (mostly "req-res"-based data exchange), though.  Will report later if any other issue comes up.

p.s. I am really enjoying using nanomsg and love its lightweight, simple, and structured design.  Thanks a lot for developing and releasing such a nice package!

**Comment by gdamore** [comment]
So is the ipc problem solved for you then or still outstanding ?

**Comment by wirebirdlabs** [comment]
> QNX has free 30-day free evaluation program

Is this particular issue #433 the one to begin the 30-day trial for the investing-developer-entity solving this issue? Or do you prefer to start this expiring timer on the next, more important issue?

I can not and will not discredit the economics behind your own concern, because I myself understand and empathize with the barriers within we must sometimes work within using proprietary platforms. but I might suggest that the burden of the economics behind whether nanomsg run on QNX is more a QNX concern than a nanomsg concern, and for this reason, this case be closed.

There does exist an issue here, no doubt, and it affects you, no argument; but this open ticket might better be on the QNX issue tracking queue, where there exists (ostensibly) allocated monetary resources and guarantees to solve such problems for subscribers and customers such as yourself.

Though, if there does not exist such budget.... then this case is still closed, and we collectively re-evaluate our decisions to rely upon proprietary "solutions" on "platforms" where money buys "guarantees".

(In the off-chance that this proprietary QNX has uncovered an odd bug that could eventually affect more mainstream and open platforms, then there exists value here to keep this issue open; though, that's not obviously, or even suspiciously, the case, based on lack of other related issue reports)

**Comment by minyang** [comment]
@gdamore I'm afraid that the IPC problem still exists with the version that I tested, AFAIK.

@wirebirdlabs Just as a clarification, I reported the issue purely based on my personal interests wishing nanomsg be more complete in terms of cross-platform support, because I like nanomsg.  Especially, its lightweight, simple, but efficiency and scalability-oriented design is what I really like.

I understand your concerns from the developer resource's perspective and you made valid points.  However, I don't think I can agree with your point where the issue is more a QNX concern than a nanomsg concern.  With the same logic, would you think it would be a MS/Windows concern if the same issue occurred on Windows, which is a commercial, mainstream, and proprietary platform?

It looks to me that nanomsg tries to support multi-platforms and I just wanted to bring the issue up to you.  Again, this issue started entirely based on my personal interests.  As I said in my previous comment, my use case is quite simple and I do not use any feature relying on IPC -- no problem with me.  Whether or not nanomsg has a plan to support QNX is totally up to nanomsg -- no question.  Also, it's totally up to you if this should be closed or not.  I'm just trying to be helpful.  On my spare time, I might try to look more into this issue if this issue remains open later.

**Comment by gdamore** [comment]
Leave the issue open.  It's real on that platform and QNX purports to be a POSIX platform.   There is clearly a problem though. 

One thought.   Perhaps the problem is the specific path chosen?

**Comment by gdamore** [comment]
Btw, are @minyang are you using IPC at all on QNX? Or just TCP.  I'd just like to get a sense about how broken Unix domain socket support on QNX things really are.

**Comment by wirebirdlabs** [comment]
> I don't think I can agree with your point where the issue is more a QNX concern than a nanomsg concern. With the same logic, would you think it would be a MS/Windows concern if the same issue occurred on Windows, which is a commercial, mainstream, and proprietary platform?

If it appears to be a problem, or even just weirdness, with the vendor's code, then it's good for the vendor to be aware and try to resolve the issue there. Now, if the vendor does not appear to be at fault, yet it's exposing some flaw in the library -- it's good for the library maintainers to have responsibility. Based on previous conversation, it felt this case fell into the former group.

That said, I appreciate the rest of your response. it's thoughtful and kind, and would agree to keep this case open for now, especially in the spirit of making nanomsg more robust. :-)

**Comment by gdamore** [comment]
Just for the record… if we eliminate “closed source/commercial operating
systems” from the supported list, or we don’t offer workarounds for some
with quirks (MacOS/Darwin!), then libnanomsg would be far less
useful/functional as a cross platform product.

While I don’t want to invest cycles *myself* in fixing QNX, since I have
neither the time nor the inclination to figure out how QNX is different, I
*do* think someone could, maybe should, address this.   Sure if QNX is
buggy somehow we should report that upstream.  But it wouldn’t be the first
platform specific bug we’ve had to workaround.  (I just yesterday
introduced a fix caused in part by Darwin returning an undocumented errno
in response to sendmsg(), for example.)

I’d love to have someone who cares about QNX step forward and offer
*modest* fixes to the problems therein.

On Thu, Oct 22, 2015 at 9:42 AM, Wirebird Labs LLC <notifications@github.com

**Comment by minyang** [comment]
As just an end-user of nanomsg, can't agree more with you @gdamore, especially in terms of usability, functionality, and cross-platform support.  Let me see if I can look into the issue a bit more, including the path issue that you suggested.

p.s. By the way, I love Apple... ;)

**Comment by minyang** [comment]
OK, I looked into a bit more yesterday.  Here are a bit of updates:

 * I got ``ipc_shutdown`` and ``separation`` working after updating IPC address from ``ipc://test.ipc`` to ``ipc:///tmp/test.ipc`` (explicit use of ``/tmp``).  Yes, @gdamore, your previous guess was correct!
 * IPC itself seems to work -- using a simple tutorial example, I was able to exchange messages via IPC protocol.  What was failing was the ``ipc`` unit test.  In``ipc``, the first part -- send/recv of 40 byte messages -- worked fine, but the following tests failed (e.g., sending large payloads (10000 bytes) one time and sending small payloads (3 bytes)).  I checked out a log file with my own debug messages, and it looks like something crashed or so (only partial messages were printed out).  From ``make check``'s point of view, this failed test looked as if it hanged.

I was too tired yesterday to narrow down which line of code cause the issue.  Hopefully, I'll have more energy tonight or this weekend.

**Comment by gdamore** [comment]
Ok, cool, this is good news.  Kind of a pity that you have to use /tmp
explicitly.   Was your build directory on a  “local” filesystem, or was it
on some kind of remote, or otherwise special, filesystem where IPC might
not be expected to work?

**Comment by minyang** [comment]
No, my build directory was just a local folder under my home, and I did all my testing on the same host.

One question, is it a normal or expected behavior where ``make check`` hangs if one of unit tests crashes (e.g., abort, exit)?  I see two different behaviors: a) hangs (in case of ``ipc``) and b) graceful handling, i.e., tests failed with coredump, moving on to the next test, and failure reports.

**Comment by gdamore** [comment]
Ideally they’d timeout and move on.  Sadly, this isn’t always done — some
just hang.  I’ve changed with Windows ctest configuration to fail hard if a
test takes over 30 sec.  I should look at doing the same for test_driver.c.

Your home directory was on local disk, right?  Strange that QNX would not
honor UNIX domain sockets there…. maybe a platform restriction?

**Comment by gdamore** [comment]
This test was a long time ago, and nanomsg has changed hugely to resolve a lot of problems and races since filed.  Submitter, if you could test against git head, or the 0.9 release (which should be released this week), it would really help.

**Comment by minyang** [comment]
Sounds like there have been lots of activities going on here.  Glad to hear that nanomsg has been updated a lot!  Sure, let me give it a shot with either git tip or Release 0.9 -- I'd be happy to try it out.

**Comment by gdamore** [comment]
Any news on this?  We have long ago released 1.0.0... absent any concrete info one way or the other, I'll probably close this.

**Comment by minyang** [comment]
This item was off my mind for a while — thanks for getting back to the issue.
Great to know that 1.0.0 has been finally released.  Please give me a few days — I’ll follow up on this during this holiday.

**Comment by gdamore** [comment]
Any news?

**Comment by minyang** [comment]
Last time I ran master on QNX (a couple months ago), QNX didn't seem to be supported by nanomsg.  I saw NN_HAVE_QNX is defined, but it wasn't actually used/implemented at that time.

With a bit of quick hacks on src and cmake (atomic add, semaphore, qnx-recognition, etc), I was able to compile the release 1.0.0 on QNX, but couldn’t get all the test apps working.  Some worked and some others didn’t -- crashed.

I couldn't follow up recent changes, but am wondering if there are some recent changes on this end?
